### PR TITLE
gtfs diff engine

### LINKED ISSUE #2: Implementation: Generic GTFS Diff Engine
### Describe the problem
There is no reusable, infrastructure-agnostic Python module for computing the difference between two GTFS datasets. Any diff logic built inside `mobility-feed-api` risks being tightly coupled to MobilityData-specific concepts (feed IDs, GCS paths, database schemas), preventing reuse across projects or publication as a standalone community tool.

### Proposed solution
Create a generic `gtfs_diff` Python package in a standalone GitHub repository. The package takes two GTFS dataset sources (URLs or local paths) and produces a structured changelog. It must have zero knowledge of MobilityData-specific concepts (no feed IDs, no GCS paths, no database).

Acceptance criteria:
- [ ] `GtfsDiff.compute()` correctly identifies added, removed, and modified rows across all supported GTFS files
- [ ] Unit tests cover: all-same feeds (empty diff), one feed empty, added rows, removed rows, modified rows, unsupported file gracefully skipped
- [ ] Package is importable locally after `pip install -e .`
- [ ] `README.md` includes install instructions and a usage example
- [ ] Zero references to MobilityData-specific infrastructure in any module

### Alternatives considered
- Inline the diff logic inside `gtfs-change-tracker`: creates tight coupling, prevents reuse as a standalone package.
- Keep the module inside `mobility-feed-api`: still couples the logic to MobilityData infrastructure and makes PyPI publication awkward.

### PR DESCRIPTION
**Pull request by cka-y**
Closes #2 

This pull request introduces the initial release of the GTFS Diff Engine, a memory-efficient Python library and CLI for comparing two GTFS feeds and producing a structured diff conforming to the GTFS Diff v2 schema. The changes include a robust implementation of the core diff logic, a clear public API, a command-line interface, detailed documentation, and supporting scripts for end-to-end usage.

The most important changes are:

**Core Functionality and API:**

* Implements the core diff logic in `engine.py`, exposing a single `diff_feeds()` function that returns a typed Pydantic model representing the diff result. [[1]](diffhunk://#diff-140eef3ba41bdcf401d507408084181f2c0ac627532b61e0f7906ea7cc926782R1-R125) [[2]](diffhunk://#diff-2a715131f975006115ec35684e89b4e459b4ad3f07907d6c83b388d63cc07722R1-R65)
* Defines the GTFS file schema, supported files, and primary key columns in `gtfs_definitions.py`, with a helper for primary key lookup.

**Command-Line Interface and Tooling:**

* Adds a Click-based CLI (`gtfs-diff`) in `cli.py`, supporting options for output file, row change cap, pretty-printing, and feed download timestamps.
* Provides a Bash script `compare_feeds.sh` to automate downloading two GTFS feeds by URL and running the diff tool, with argument parsing and error handling.

**Documentation and Examples:**

* Expands `README.md` with a comprehensive overview, installation instructions, usage examples, API reference, supported files table, output schema example, and implementation notes on memory efficiency.
* Adds `docs/architecture.md` detailing design goals, module structure, streaming diff algorithm, edge case handling, and future improvements.

**Packaging and Project Setup:**

* Configures packaging with `pyproject.toml` for installation, development, and test dependencies, and sets up the CLI entry point.
* Adds package versioning and module entry points in `__init__.py` and `__main__.py`. [[1]](diffhunk://#diff-b3e841ae539867aff26a206b0697d7a03c67e5f305bd709437e273c420b3082eR1-R3) [[2]](diffhunk://#diff-3923a55a5a4b2ff30d4a137bbd3b34cdeae4e1634a68be005a514ff2ef2872fcR1-R6)

These changes collectively deliver a ready-to-use, well-documented GTFS diff engine suitable for both programmatic and CLI-based workflows.

### CODE REVIEW

**Comment by jcpitre** [comment]
Some testing:

- Went to the extreme and tested it on the biggest feed we have: 
  - mdb-2014-202603090029.zip
  - mdb-2014-202603110034.zip
  
Each is about 1.5GB zipped and over 38 miilion entries in shapes.txt and 50 million of entries in stop_times.txt
I had to stop it as it was working on shapes.txt, using 4GB of memory:

```
[gtfs-diff 16:31:57 4321MB]   [shapes.txt] scan done in 0.0s — added=50,118,266 deleted=49,537,043 modified=0
```
Which indicates that there was 50+ million addition and about as many deletion. 

Examining shapes.txt, we can see that they use some uuid for the shape_id, e.g.:
```
0000abe0-5266-475b-808a-5cf929120a80,50.118257595,-5.540823891,229,
```
That seems to be completely regenerated for each version. That means that trying to do a diff in that kind of dataset is not possible.

- We should make a survey of the dataset we have to see if it's common to have regenerated ids.
- We should consider finding some heuristics that would tell us quickly if a there is a possibility of meaningful comparison between 2 datasets.

**Comment by jcpitre** [comment]
Tested with STM: mdb-2126-202511111837.zip vs mdb-2126-202511130041.zip.
I spot checked some of the reported differences, and they seems ok, except:

Problems:
- In shapes.txt and stops.txt, there was a significant number (thousands?) of modified lines that were due to coordinates having extra zeros:
```
          {
            "identifier": {
              "shape_id": "11071",
              "shape_pt_sequence": "150001"
            },
            "raw_value": "11071,45.518332,-73.556250,150001",
            "base_line_number": 35,
            "new_line_number": 35,
            "field_changes": [
              {
                "field": "shape_pt_lon",
                "base_value": "-73.55625",
                "new_value": "-73.556250"
              }
            ]
          }
```

- In feed_info.txt, the base file had an empty line at the end, and this translated to this jason entry:
```
       "deleted": [
          {
            "identifier": {
              "feed_publisher_name": "",
              "feed_publisher_url": "",
              "feed_lang": "",
              "feed_start_date": "",
              "feed_end_date": "",
              "feed_version": ""
            },
            "raw_value": ",,,,,",
            "base_line_number": 3
          },
```

**Comment by jcpitre** [comment]
Modified code to ignore extra zeros in coordinates.

**Comment by emmambd** [comment]
@jcpitre Were the big feeds DELFI? I'm curious which datasets they were. IDFM told us their data is 120MB zipped.

**Comment by jcpitre** [comment]
> @jcpitre Were the big feeds DELFI? I'm curious which datasets they were. IDFM told us their data is 120MB zipped.

mdb-2014 is the UK aggregate feed. Size is about 1.5GB zipped. Here are the datasets I used:

- [mdb-2014-202603090029.zip](https://storage.googleapis.com/mobilitydata-datasets-prod/mdb-2014/mdb-2014-202603090029/mdb-2014-202603090029.zip)
- [mdb-2014-202603110034.zip](https://storage.googleapis.com/mobilitydata-datasets-prod/mdb-2014/mdb-2014-202603110034/mdb-2014-202603110034.zip)

File: scripts/compare_feeds.sh [hunk 1]
```diff
@@ -0,0 +1,126 @@
+#!/usr/bin/env bash
+# compare_feeds.sh — Download two GTFS feeds by URL and diff them.
+#
+# Usage:
+#   ./scripts/compare_feeds.sh <BASE_URL> <NEW_URL> [OPTIONS]
```
Comment (davidgamez) [inline scripts/compare_feeds.sh:5]: [suggestion]: It would be nice if the shell script also supported local zip and folder options. _This can be done outside this PR._

File: src/gtfs_diff/engine.py [hunk 2]
```diff
@@ -0,0 +1,613 @@
...
+    try:
+        row = next(reader)
+    except StopIteration:
+        return {col: "" for col in headers}
+    if len(row) < len(headers):
+        row = row + [""] * (len(headers) - len(row))
+    return dict(zip(headers, row))
+
+
+def _values_differ(a: str, b: str) -> bool:
```
Comment (davidgamez) [inline src/gtfs_diff/engine.py:163]: [question]: Do we need to trim the parameters to make sure only the "actual" content is compared?

File: src/gtfs_diff/engine.py [hunk 3]
```diff
@@ -0,0 +1,613 @@
...
+        raise ValueError(
+            f"Unsupported feed path: {path!r}.  Must be a .zip file or a directory."
+        )
+
+
+# ---------------------------------------------------------------------------
+# Per-file diff
+# ---------------------------------------------------------------------------
+
+def _diff_file(
```
Comment (davidgamez) [inline src/gtfs_diff/engine.py:252]: [suggestion]: I find this function long. To improve readability, I suggest splitting it into multiple functions. We can naturally split it "per section", as it's already commented within the function.

File: scripts/compare_feeds.sh (hunk 1 above)
Comment (jcpitre) [reply scripts/compare_feeds.sh:5]: Added an issue on this: #6

File: src/gtfs_diff/engine.py (hunk 2 above)
Comment (jcpitre) [reply src/gtfs_diff/engine.py:163]: Trimmed

File: src/gtfs_diff/engine.py (hunk 3 above)
Comment (jcpitre) [reply src/gtfs_diff/engine.py:252]: Did some refactoring.

File: src/gtfs_diff/engine.py [hunk 4]
```diff
@@ -0,0 +1,655 @@
...
+
+    String-identical values are equal (fast path).
+    If they differ as strings, attempt numeric comparison — this silently
+    ignores cosmetic differences like trailing zeros in coordinate fields
+    (e.g. '-73.55625' vs '-73.556250').
+    Non-numeric strings fall back to string equality.
+    """
+    a = a.strip(),
+    b = b.strip()
+    if a == b:
```
Comment (davidgamez) [inline src/gtfs_diff/engine.py]: We should do no case-sensitive comparison here
```suggestion
    if a.lower() == b.lower():
```

**Review by davidgamez — APPROVED** [review]

File: src/gtfs_diff/engine.py (hunk 4 above)
Comment (jcpitre) [reply src/gtfs_diff/engine.py]: This will have to be revisited eventually. As you mentioned @davidgamez maybe some fields can be compared case insensitive, and others have to be compared case sensitive. 
As an example, one of the warnings of the validator is  [mixed_case_recommended_field](https://gtfs-validator.mobilitydata.org/rules.html#mixed_case_recommended_field-rule). If a given field changes case between 2 datasets, it could result in a mixed_case_recommended_field warning in the validation report of one dataset and not in the other. So there is a difference, but the comparator would not say anything about it, which would be a discrepency.
